### PR TITLE
feat: add --strict flag for partial spec acceptance with warnings

### LINKED ISSUE #11: feat: Support partial OpenAPI spec acceptance with warnings for unsupported content types
## Problem

Currently, when a user tries to add an OpenAPI spec that contains endpoints with unsupported content types (like `multipart/form-data`), the entire spec is rejected with a validation error. This is overly restrictive for APIs that have mostly JSON endpoints but a few file upload endpoints.

Example error:
```
Validation Error
Unsupported request body content type 'multipart/form-data' in post /channels/{channel_name}/artifacts. Only 'application/json' is supported in v1.0.
```

## Proposed Solution

Implement partial spec acceptance that:

1. **Accepts specs with warnings** - Skip unsupported endpoints instead of rejecting the entire spec
2. **Shows clear warnings** during `config add` about which endpoints were omitted
3. **Adds `--strict` flag** to maintain current behavior for users who want strict validation

### Example Warning Output
```
Warning: Skipping 3 endpoints with unsupported content types:
- POST /api/upload (multipart/form-data) - file uploads not supported
- PUT /api/config (text/plain) - only JSON request bodies supported  
- POST /api/legacy (application/xml) - XML content not supported

23 of 26 endpoints will be available. Use --strict to reject specs with unsupported content types.
```

### Binary File Types (skip with warning)
- `multipart/form-data`
- `application/octet-stream`
- `image/*`
- `application/pdf`

### Alternative Text Formats (skip with warning)
- `application/xml`, `text/xml`
- `application/x-www-form-urlencoded`
- `text/plain`
- `text/csv`

### JSON-Compatible (skip with warning, consider for future)
- `application/x-ndjson`
- `application/graphql`

## Implementation Notes

1. Modify `SpecValidator` to collect warnings instead of failing immediately
2. Filter unsupported endpoints during spec transformation/caching
3. Add `--strict` flag to `ConfigAddArgs` in CLI
4. Store metadata about skipped endpoints for `config list --verbose`

## Benefits

- Better UX: Users can use most of an API even if some endpoints aren't supported
- Progressive enhancement: When new content types are supported, endpoints automatically become available
- Clear communication: Users understand exactly what's not supported
- Practical: Many real-world APIs mix JSON endpoints with a few file upload endpoints

## Related

- Original discussion about multipart/form-data limitation
- Aligns with Aperture's JSON-first design philosophy while being more user-friendly

### PR DESCRIPTION
**Pull request by kioku**
## Summary
- Implements partial OpenAPI spec acceptance for unsupported content types
- Adds `--strict` flag to `config add` command
- Closes #11

## Changes
- Add `ValidationResult` struct to collect warnings alongside errors
- Add `--strict` flag to `config add` command (default: non-strict)
- In non-strict mode, skip endpoints with unsupported content types
- Display detailed warnings for skipped endpoints
- Filter skipped endpoints from cached spec during transformation
- Add comprehensive integration tests for strict/non-strict modes
- Update architecture docs to explain validation modes

### Non-strict mode (default)
```bash
$ aperture config add my-api ./spec-with-multipart.yaml

Warning: Skipping 2 endpoints with unsupported content types:
  - POST /users/{userId}/avatar (multipart/form-data) - content type 'multipart/form-data' is not supported
  - POST /documents (multipart/form-data) - content type 'multipart/form-data' is not supported

Use --strict to reject specs with unsupported content types.
```

### Strict mode
```bash
$ aperture config add --strict my-api ./spec-with-multipart.yaml

Error: Unsupported request body content type 'multipart/form-data' in POST /users/{userId}/avatar. Only 'application/json' is supported in v1.0.
```

## Test Plan
- [x] Unit tests for new `ValidationResult` and `validate_with_mode` API
- [x] Integration tests for CLI behavior in strict/non-strict modes
- [x] Integration tests verifying multipart endpoints are excluded from generated commands
- [x] All existing tests pass with `strict=true` for backward compatibility

### CODE REVIEW

**Comment by kioku** [comment]
## Code Review: Strict Flag Implementation

I've conducted a thorough review of the `--strict` flag implementation. The core functionality is well-designed and achieves its goal of allowing partial spec acceptance. However, there are some critical issues that need to be addressed before merging.

### Critical Issues

1. **Broken Test Method** (`src/config/manager.rs:562-567`)
   - The test-only method `add_spec_from_url_with_timeout` still uses the old validation API
   - This needs to be updated to use `validate_with_mode()` and pass skip_endpoints to `transform()`
   ```rust
   // Current (broken)
   validator.validate(&openapi_spec)?;
   transformer.transform(name, &openapi_spec)?;
   
   // Should be
   let validation_result = validator.validate_with_mode(&openapi_spec, strict);
   if \!validation_result.is_valid() {
       return validation_result.into_result();
   }
   let cached_spec = transformer.transform(name, &openapi_spec, &validation_result.skip_endpoints)?;
   ```

2. **Inconsistent Error Types** (`src/config/manager.rs:585-586`)
   - Test method uses `Error::Config` while production code uses `Error::SerializationError`
   - This inconsistency could lead to different error handling paths

### Important Issues

3. **Case Sensitivity Inconsistency** (`src/spec/transformer.rs:79-81`)
   - Path comparison is case-sensitive while method comparison is case-insensitive
   - Consider normalizing both for consistent behavior

4. **Missing Architecture Documentation**
   - PR description mentions ADR-003, but no new ADR was created
   - Existing ADR-003 is about "test-isolation-base-url-configuration"

### What's Working Well

- Excellent backward compatibility with default `strict=false`
- Comprehensive integration test coverage
- Clear and helpful warning messages
- Proper endpoint filtering from cached specs
- Good separation of concerns in the implementation

### Minor Suggestions

- Consider adding tests for edge cases (multiple content types, mixed content in same endpoint)
- Warning display format could use color-coding or better structure for readability
- Documentation comments could include more specific examples

### Summary

The implementation successfully enables partial OpenAPI spec acceptance with clear warnings, which will be valuable for users working with APIs that have mixed endpoint support. Once the critical issues (especially the broken test method) are fixed, this will be ready to merge.

Great work on maintaining backward compatibility and providing a clean API\!

**Review by kioku — COMMENTED** [review]
## PR Review: Partial OpenAPI Spec Acceptance

Thank you for implementing this feature\! The core functionality works well, but I've identified several issues that need attention before merging.

### What Works Well
- Clean implementation using `ValidationResult` struct
- Good test coverage for multipart/form-data
- Well-documented with ADR-006
- Maintains backward compatibility with `--strict` flag

### Critical Issues

#### 1. Path Comparison Bug (HIGH SEVERITY)
In `src/spec/transformer.rs:79-82`:
```rust
let should_skip = skip_endpoints.iter().any(|(skip_path, skip_method)| {
    skip_path.eq_ignore_ascii_case(path)
        && skip_method.eq_ignore_ascii_case(method)
});
```
**Problem**: OpenAPI paths are case-sensitive. Using `eq_ignore_ascii_case` for paths could incorrectly filter endpoints.
- Example: `/Users` and `/users` are different endpoints in OpenAPI
- This could cause valid endpoints to be accidentally skipped

**Fix**: Use exact comparison for paths:
```rust
skip_path == path && skip_method.eq_ignore_ascii_case(method)
```

#### 2. Incomplete Content Type Support (MEDIUM SEVERITY)
Issue #11 lists many unsupported content types, but the implementation only handles `multipart/form-data`:
- Missing: `application/xml`, `text/xml`
- Missing: `application/x-www-form-urlencoded`
- Missing: `text/plain`, `text/csv`
- Missing: `application/octet-stream`
- Missing: `image/*`, `application/pdf`

These content types will still cause rejection even in non-strict mode, which doesn't fully address the issue.

#### 3. Lost Warning Information (MEDIUM SEVERITY)
Warnings are only displayed to stderr during `config add`. There's no way to retrieve which endpoints were skipped later. Consider:
- Storing skipped endpoints in spec metadata
- Showing them in `config list --verbose`

### Minor Issues

1. **Misleading Commit Message**: Commit 34ec99d says "default true for backward compatibility" but the implementation defaults to false (non-strict)

2. **Missing Test Coverage**:
   - No tests for other content types (XML, form-urlencoded, etc.)
   - No tests for endpoints with multiple content types
   - No edge cases for malformed content types

### Recommendations

1. Fix the path comparison to be case-sensitive
2. Either:
   - Extend the implementation to handle all content types from issue #11, or
   - Document which content types are supported in non-strict mode
3. Consider persisting warning information for later retrieval
4. Add test coverage for other content types
5. Update the `--strict` flag help text to be more descriptive

The PR implements the core feature well, but these issues should be addressed to fully resolve issue #11 and prevent potential bugs.

**Review by kioku — COMMENTED** [review]
## Critical PR Review: Add --strict flag for partial spec acceptance

After thorough analysis of the implementation, I've identified several critical issues that must be addressed before merging:

### CRITICAL ISSUES

#### 1. Content Type Matching is Too Restrictive (HIGH SEVERITY)

The implementation at `src/spec/validator.rs:340` uses exact string comparison that will incorrectly reject valid JSON content types:

```rust
.filter(|ct| *ct \!= "application/json")
```

This will reject:
- `APPLICATION/JSON` (case variations)
- `application/json; charset=utf-8` (with parameters)
- `  application/json  ` (with whitespace)

The test at `tests/content_type_validation_tests.rs:362` expects this restrictive behavior, but this contradicts real-world HTTP content type handling where media types are case-insensitive per RFC 7231.

**Required Fix:**
```rust
let content_type_base = ct.split(';').next().unwrap_or(ct).trim();
content_type_base.eq_ignore_ascii_case("application/json")
```

#### 2. Cache Format Version Not Incremented (HIGH SEVERITY)

The `skipped_endpoints` field was added to `CachedSpec` but `CACHE_FORMAT_VERSION` remains at 1 (`src/cache/models.rs:32`). This will cause deserialization errors for users with existing caches.

**Required Fix:** Increment `CACHE_FORMAT_VERSION` to 2 and add migration logic or clear error messaging.

### IMPORTANT ISSUES

#### 3. Inconsistent Default Behavior in Tests (MEDIUM SEVERITY)

The test helper `add_spec_from_url_with_timeout` at `src/config/manager.rs:615` defaults to `strict=true`, contradicting the new default non-strict behavior. This inconsistency could lead to confusing test failures.

#### 4. Missing Deprecation of Old API (MEDIUM SEVERITY)

The original `validate()` method still exists and defaults to strict mode. This creates API confusion since the new pattern is `validate_with_mode()`. Consider deprecating the old method with clear migration guidance.

### MISSING TEST COVERAGE

1. **Content type edge cases**: No tests for `application/json; charset=utf-8` or other valid JSON content type variations
2. **Cache migration**: No tests for handling old cache format without `skipped_endpoints` field
3. **Backward compatibility**: No tests ensuring `validate()` maintains strict behavior for existing code

### POSITIVE ASPECTS

The implementation does several things well:
- Smart endpoint handling: Endpoints with both JSON and unsupported types are correctly included
- Clear warning messages with user-friendly explanations
- Proper filtering ensures skipped endpoints don't appear at runtime
- Comprehensive integration tests for the main functionality
- Well-documented ADR explaining the design decision

### RECOMMENDATIONS

1. **Fix content type matching** to handle real-world HTTP content types properly
2. **Increment cache version** and add migration handling
3. **Add comprehensive tests** for content type edge cases
4. **Deprecate old validate() method** with `#[deprecated]` attribute
5. **Update ADR** to document case sensitivity behavior

### IMPACT ASSESSMENT

- **User Impact**: HIGH - Current implementation will reject many valid OpenAPI specs with standard JSON content types
- **Migration Risk**: HIGH - Cache version mismatch will break existing installations
- **API Stability**: MEDIUM - Inconsistent defaults between validation methods

### VERDICT

This PR should **NOT be merged** until at least issues #1 (content type matching) and #2 (cache version) are fixed. These are blocking issues that will cause significant problems for users.

The overall design and approach are sound, but these implementation details need correction before the feature can be safely released.

File: src/spec/validator.rs [hunk 1]
```diff
@@ -204,22 +328,83 @@ impl SpecValidator {
...
+                    ValidationWarning {
+                        endpoint: UnsupportedEndpoint {
+                            path: path.to_string(),
+                            method: method.to_uppercase(),
+                            content_type: content_types.join(", "),
+                        },
+                        reason: "endpoint has no supported content types".to_string(),
+                    }
+                };
+                result.add_warning(warning);
```
Comment (kioku) [inline src/spec/validator.rs:405]: This code is hard to read given the number of nested statements. It should be refactored and flattened.
